src/logcollector/read_journald.c

```diff
@@ -0,0 +1,106 @@
+#ifdef HAVE_SYSTEMD
+#include "shared.h"
+#include "logcollector.h"
+#include <systemd/sd-journal.h>
+
+int prime_sd_journal(sd_journal **jrn) {
+  int ret;
+  ret = sd_journal_open(jrn, SD_JOURNAL_LOCAL_ONLY);
+  if (ret < 0) {
+    merror("%s: ERROR: Unable to open journal", ARGV0);
+    return ret;
+  }
+  ret = sd_journal_get_fd(*(jrn));
+  if (ret < 0) {
+    merror("%s: ERROR: Unable to get journal fd", ARGV0);
+    return ret;
+  }
+  ret = sd_journal_seek_tail(*(jrn));
+  if (ret < 0) {
+    merror("%s: ERROR: Unable to seek journal tail", ARGV0);
+    return ret;
+  }
+  // prime sd_journal_next before the reader loop
+  ret = sd_journal_previous(*(jrn));
+  if (ret < 0) {
+    merror("%s: ERROR: Unable to seek journal previous", ARGV0);
+  }
+  return ret;
+}
+
+void *sd_read_journal(__attribute__((unused)) char *unit) {
+  sd_journal *jrn;
+  int ret;
+  const char *jmsg, *jsrc, *jhst;
+  size_t len;
+  struct timeval tv;
+  uint64_t curr_timestamp;
+  time_t nowtime;
+  struct tm *nowtm;
+  char tmbuf[64], final_msg[OS_MAXSTR];
+
+  ret = prime_sd_journal(&jrn);
+  if (ret < 0) {
+    merror("%s: ERROR: Unable to prime journal", ARGV0);
+    return NULL;
+  }
+  ret = sd_journal_next(jrn);
+  // Anything negative is an error
+  while (ret >= 0) {
+    // No messages, wait for data
+    if (ret == 0) {
+      while ((sd_journal_wait(jrn, (uint64_t)-1)) == SD_JOURNAL_NOP) {
+        sleep(1);
+      }
+    } else {
+      // Check if data is coming from the unit starting with our passed selector
+      ret = sd_journal_get_data(
+        jrn,
+        "SYSLOG_IDENTIFIER",
+        (const void **)&jsrc,
+        &len
+      );
+      // Strip off "SYSLOG_IDENTIFIER=" prefix for unit comparison inline
+      if (strstr((char *)(jsrc + 18), unit) == (char *)(jsrc + 18) ) {
+        // Read hostname
```

Comment:
There should be a way to log all journald stuff as reading /var/log/messages as this just logs some units output.